### PR TITLE
Abstract namespace addresses support for Unix domain sockets

### PR DESCRIPTION
**Pull request by vtikoo**
Conventionally Unix domain socket addresses are a path in the file hierarchy. Abstract namespace identifiers allow addresses without creating a corresponding file path. They are differentiated from conventional UDS addresses by starting the address with a null character.

Pros:
- Cleaning up for socket file path is not required.
- User does not need permissions on the directories in the path.

Cons:
- No access control by default: Unix file permissions do not carry over to Abstract namespace identifiers.
- Portability: Linux only feature.

Signed-off-by: Vikas Tikoo <vikasamar@gmail.com>

### CODE REVIEW

File: kernel/syscall.c [hunk 1]
```diff
@@ -387,17 +387,36 @@ static int _socketaddr_to_str(
...
-    const uint8_t ip3 = p[4];
-    const uint8_t ip4 = p[5];
-
-    if (snprintf(out, limit, "%u.%u.%u.%u:%u", ip1, ip2, ip3, ip4, port) >=
-        (int)limit)
+    if (addr->sa_family == AF_UNIX)
     {
-        ERAISE(-ENAMETOOLONG);
+        if (addr->sa_data[0] &&
+            myst_snprintf(out, limit, "%s", addr->sa_data) == ERANGE)
```
Comment (jxyang) [inline kernel/syscall.c:393]: Should it be `-ERANGE`?

File: kernel/syscall.c [hunk 2]
```diff
@@ -387,17 +387,36 @@ static int _socketaddr_to_str(
...
+        {
+            ERAISE(-ENAMETOOLONG);
+        }
+        // abstract namespace address
+        else if (
+            myst_snprintf(
+                out, limit, "%s(abstract namespace)", addr->sa_data + 1) ==
+            ERANGE)
```
Comment (jxyang) [inline kernel/syscall.c:401]: Likewise

File: kernel/syscall.c [hunk 3]
```diff
@@ -387,17 +387,36 @@ static int _socketaddr_to_str(
...
-        ERAISE(-ENAMETOOLONG);
+        if (addr->sa_data[0] &&
+            myst_snprintf(out, limit, "%s", addr->sa_data) == ERANGE)
+        {
+            ERAISE(-ENAMETOOLONG);
+        }
+        // abstract namespace address
+        else if (
+            myst_snprintf(
+                out, limit, "%s(abstract namespace)", addr->sa_data + 1) ==
```
Comment (jxyang) [inline kernel/syscall.c:400]: maybe just "abstract" or "abs ns" to reduce the chance of overflow.

File: solutions/cpython-tests/Makefile [hunk 4]
```diff
@@ -109,6 +109,14 @@ To enable python source listing, do \n\
 before launching gdb.\033[0m\n')" \
            --args $(MYST_EXEC) $(OPTS) $(FS) /cpython/python -m unittest Lib.test.$(TESTCASE) -v
 
+testcase-mpdb:
+	killall myst 2> /dev/null || echo ""
```
Comment (jxyang) [inline solutions/cpython-tests/Makefile:113]: Is this related?

File: tests/sockets/sockets.c [hunk 5]
```diff
@@ -260,13 +281,16 @@ void test_sockets(args_t* args)
     pthread_t srv_thread;
     pthread_t cli_thread;
 
+    pthread_barrier_init(&bar, NULL, 2);
+
     assert(pthread_create(&srv_thread, NULL, _srv_thread_func, args) == 0);
-    _sleep_msec(100);
```
Comment (jxyang) [inline tests/sockets/sockets.c:264]: Why the sleep is no longer needed?

Comment (vtikoo) [reply tests/sockets/sockets.c:264]: I changed the test to use a `pthread_barrier_t`.

File: solutions/cpython-tests/Makefile (hunk 4 above)
Comment (vtikoo) [reply solutions/cpython-tests/Makefile:113]: This was useful in debugging cpython stacktraces. The related test `test_multiprocessing_spawn` is not part of this PR, being blocked by shared memory support.

File: kernel/syscall.c (hunk 1 above)
Comment (vtikoo) [reply kernel/syscall.c:393]: Thanks for catching this!

**Review by jxyang — APPROVED** [review]
